futures-util/src/future/abortable.rs

```diff
@@ -0,0 +1,131 @@
+use futures_core::future::Future;
+use futures_core::task::{self, AtomicWaker, Poll};
+use std::marker::Unpin;
+use std::mem::PinMut;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicBool, Ordering};
+
+/// A future which can be remotely short-circuited using an `AbortHandle`.
+#[derive(Debug, Clone)]
+#[must_use = "futures do nothing unless polled"]
+pub struct Abortable<T> {
+    future: T,
+    inner: Arc<AbortInner>,
+}
+
+impl<T: Unpin> Unpin for Abortable<T> {}
+
+impl<T> Abortable<T> where T: Future {
+    /// Creates a new `Abortable` future using an existing `AbortRegistration`.
+    /// `AbortRegistration`s can be acquired through `AbortHandle::new`.
+    ///
+    /// When `abort` is called on the handle tied to `reg` or if `abort` has
+    /// already been called, the future will complete immediately without making
+    /// any further progress.
+    pub fn new(future: T, reg: AbortRegistration) -> Self {
```

Comment:
Probably should be private

Reply:
Ah no, you intend to make in to possible to create an `AbortHandle` without creating an `Abortable` future. I'm not against adding this functionality, but it seems a bit over-engineered. Is there an actual use case for this?

*Anywho*, there should probably be a unit test and a quick code example in the docs if this functionality should exist.

Reply:
@cramertj I've found your old comment where you explain why you want this. I now fully agree that we want this functionality.

https://github.com/rust-lang-nursery/futures-rs/pull/696#issuecomment-356692685